### PR TITLE
My code review.

### PR DESCRIPTION
**Pull request by pranavm7**
Analysis of program: Everything builds properly. Test coverage is high.

Was the program available in UC Github on time: Yes (I had issues with pushing but DevOps responded to me quickly and we got it resolved.)

Is the program documented/commented well enough for you to understand: 
Code was well written that minimizes the need for comments, although it was lacking in documentation which I added in my pull request.

Does the program compile: Yes

Rationale behind your changes: 
- Added missing documentation.
- Added a conditional to account for data not being retrieved from the database/backend.
- Refactored function to be more readable and enforced access control.

What I learned from this code review:

- I learned how to use a MutableMap
- I learned how to check for a successful HTTP request
- I learned documentation using KDoc

Commits to my project:
- [Fixed CircleCI](https://github.com/pranavm7/UniConnect/commit/d6d8ba46fbaac5686bf5edb9c63813bcf91832af)
- [Added CircleCI config](https://github.com/pranavm7/UniConnect/commit/80fdcf25381589687731acc0c8fada50c51e167b)
- [Created a diagram for README.md](https://github.com/pranavm7/UniConnect/commit/59214c8e04771f8f49cb2967effe9f44cf410258)
- There are additional commits on the [repository](https://github.com/pranavm7/UniConnect)

### CODE REVIEW

File: app/src/main/java/com/cookit/RetrofitClientInstance.kt [hunk 1]
```diff
@@ -5,13 +5,13 @@ import retrofit2.converter.gson.GsonConverterFactory
 
 object RetrofitClientInstance {
     private var retrofit: Retrofit? = null
-    private val BASE_URL = ""
+    private const val BASE_URL = ""
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/RetrofitClientInstance.kt:8]: As it follows the naming convention of a constant, I thought adding the const keyword made sense.

File: app/src/main/java/com/cookit/RetrofitClientInstance.kt [hunk 2]
```diff
@@ -5,13 +5,13 @@ import retrofit2.converter.gson.GsonConverterFactory
...
     // data used for testing located in RecipeServiceStub class
 
     val retrofitInstance : Retrofit?
         get() {
             if (retrofit == null) {
-                retrofit = retrofit2.Retrofit.Builder()
+                retrofit = Retrofit.Builder()
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/RetrofitClientInstance.kt:14]: As `Retrofit` was already imported from retrofit2 at the top of the file, mentioning it here was redundant.

File: app/src/main/java/com/cookit/dao/IRecipeDAO.kt [hunk 3]
```diff
@@ -6,5 +6,5 @@ import retrofit2.http.GET
 
 interface IRecipeDAO {
     @GET("")
-    fun getAllRecipes() : Call<Set<Recipe>>
+    suspend fun getAllRecipes() : Call<Set<Recipe>>
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/dao/IRecipeDAO.kt:9]: added the suspend function to enable asynchronous calls

File: app/src/main/java/com/cookit/service/RecipeService.kt [hunk 4]
```diff
@@ -8,14 +8,27 @@ import kotlinx.coroutines.async
...
+ *  A service class related to Recipes.
+ */
 class RecipeService {
+    /**
+     * Gets Recipe from the database.
+     *
+     * This function implements the [IRecipeDAO] interface to fetch recipe data from the specified url.
+     * @throws Exception Throws an exception when the server response is not successful.
+     * @return Response body from the url specified at [IRecipeDAO.getAllRecipes] GET wrapper
+     */
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/service/RecipeService.kt:21]: Added documentation for service class following [KDoc convention](https://kotlinlang.org/docs/kotlin-doc.html#kdoc-syntax)

File: app/src/main/java/com/cookit/service/RecipeService.kt [hunk 5]
```diff
@@ -8,14 +8,27 @@ import kotlinx.coroutines.async
...
-            val recipes = async { service?.getAllRecipes() }
-            var result = recipes.await()?.awaitResponse()?.body()
-            return@withContext result
+            val response = service!!.getAllRecipes().awaitResponse()
+            if (response.isSuccessful)
+            {
+                return@withContext response.body()
+            }else{
+                throw Exception("Failed to get recipes. Server Response: ${response.code()}")
+            }
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/service/RecipeService.kt:31]: Added a check constraint to ensure that HTTP request was successful (Response Code: 200)

File: app/src/main/java/com/cookit/service/RecipeServiceStub.kt [hunk 6]
```diff
@@ -2,71 +2,80 @@ package com.cookit.service
...
 import com.cookit.dto.Recipe
 
-class RecipeServiceStub {
+/**
+ *  A service class to simulate data.
+ *
+ *  This class is a test alternative to the [RecipeService] class.
+ *  It generates data to be used during testing using [fetchRecipes].
+ */
+internal class RecipeServiceStub {
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/service/RecipeServiceStub.kt:11]: Added access control

Comment (pranavm7) [inline app/src/main/java/com/cookit/service/RecipeServiceStub.kt:11]: Added documentation according to KDoc

File: app/src/main/java/com/cookit/service/RecipeServiceStub.kt [hunk 7]
```diff
@@ -2,71 +2,80 @@ package com.cookit.service
...
+     */
+    fun fetchRecipes(): Set<Recipe> {
 
-    suspend fun fetchRecipes(): Set<Recipe>? {
-
-        var fakeRecipe = mutableSetOf<Recipe>()
-
-         fakeRecipe.add(Recipe(
+        return mutableSetOf<Recipe>(
+          Recipe(
```
Comment (pranavm7) [inline app/src/main/java/com/cookit/service/RecipeServiceStub.kt:22]: removed async as it was not necessary, also consolidated the var declaration directly to the return statement.

File: app/src/main/java/com/cookit/RetrofitClientInstance.kt (hunk 2 above)
Comment (discospiff) [reply app/src/main/java/com/cookit/RetrofitClientInstance.kt:14]: Good.

File: app/src/main/java/com/cookit/RetrofitClientInstance.kt (hunk 1 above)
Comment (discospiff) [reply app/src/main/java/com/cookit/RetrofitClientInstance.kt:8]: Good.

File: app/src/main/java/com/cookit/service/RecipeService.kt [hunk 8]
```diff
@@ -8,6 +8,14 @@ import kotlinx.coroutines.async
...
 import retrofit2.awaitResponse
 
+/**
+ * A service class related to Recipes
+ * Gets Recipe from the database.
+ *
+ * This function implements the [IRecipeDAO] interface to fetch recipe data from the specified url.
+ * @throws Exception Throws an exception when the server response is not successful.
+ * @return Response body from the url specified at [IRecipeDAO.getAllRecipes] GET wrapper
+ */
```
Comment (discospiff) [inline app/src/main/java/com/cookit/service/RecipeService.kt:18]: Ok, good.  KDoc that is descriptive.

File: app/src/main/java/com/cookit/service/RecipeServiceStub.kt [hunk 9]
```diff
@@ -2,71 +2,80 @@ package com.cookit.service
...
+ */
+internal class RecipeServiceStub {
 
-    // service stub of simulated data used in testing
+    /**
+     * Alternative to [RecipeService.fetchRecipes]
+     *
+     * Generates sample data to be used during testing.
+     * @return A sample dataset.
+     */
```
Comment (discospiff) [inline app/src/main/java/com/cookit/service/RecipeServiceStub.kt:18]: This is the proper way to document a method. 

Inline comments (//) should be within a method; not above.

File: app/src/main/java/com/cookit/service/RecipeServiceStub.kt (hunk 7 above)
Comment (discospiff) [reply app/src/main/java/com/cookit/service/RecipeServiceStub.kt:22]: Should this be mutableStateOf<List<Recipe>> if it is returning multiple?

File: app/src/test/java/com/cookit/RecipeServiceIntegrationTest.kt [hunk 10]
```diff
@@ -15,13 +16,14 @@ import kotlinx.coroutines.test.runTest
...
+@ExperimentalCoroutinesApi
 class RecipeServiceIntegrationTest {
 
     @get:Rule
     var rule: TestRule = InstantTaskExecutorRule()
 
-    lateinit var recipeService : RecipeServiceStub
-    var allRecipes : Set<Recipe>? = HashSet<Recipe>()
+    private lateinit var recipeService : RecipeServiceStub
+    private var allRecipes : Set<Recipe>? = HashSet<Recipe>()
```
Comment (discospiff) [inline app/src/test/java/com/cookit/RecipeServiceIntegrationTest.kt:26]: Good.

**Comment by discospiff** [comment]
Several good suggestions here... and it looks like the group has already merged them!

**Pull request merged**